### PR TITLE
Skip updating signature-help if there's another update request in the update-queue.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
@Pilchie Can you try this with your scenario where you saw SignatureHelp slowdowns?

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session.cs [hunk 1]
```diff
@@ -9,6 +9,15 @@ internal partial class Controller
...
         internal partial class Session : Session<Controller, Model, ISignatureHelpPresenterSession>
         {
+            /// <summary>
+            /// When ther user moves the caret we issue retrigger commands.  There may be a long
+            /// chain of these, and they may take time for each to process.  This can be visible 
+            /// to the user as a long delay before the signature help items update.  To avoid this
+            /// we keep track if there is new outstanding retrigger command and we bail on the
+            /// computation if another is in the queue.
+            /// </summary>
+            private int _retriggerId;
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session.cs]: Does this need to be marked volatile? I can never remember.

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs [hunk 2]
```diff
@@ -68,30 +76,36 @@ internal partial class Session
 
                         if (triggerInfo.TriggerReason == SignatureHelpTriggerReason.RetriggerCommand)
                         {
-                            if (currentModel == null ||
-                                (triggerInfo.TriggerCharacter.HasValue && !currentModel.Provider.IsRetriggerCharacter(triggerInfo.TriggerCharacter.Value)))
+                            if (currentModel == null)
+                            {
+                                return currentModel;
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs]: Also known as "null".

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs [hunk 3]
```diff
@@ -158,17 +172,18 @@ private static SignatureHelpItem GetSelectedItem(Model currentModel, SignatureHe
...
-            private async Task<Tuple<ISignatureHelpProvider, SignatureHelpItems>> ComputeItemsAsync(
-                IList<ISignatureHelpProvider> providers,
+                => p1.ToString() == p2.ToString();
+
+            /// <summary>
+            /// Returns <code>null</code> if our work was preempted and we want to return the 
+            /// previous model we've computed.
+            /// </summary>
+            private async Task<(ISignatureHelpProvider provider, SignatureHelpItems items)?> ComputeItemsAsync(
+                int id,
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs]: Can we call this retriggerId or something so it's clear it relates to the other field? `id` is a bit vague.

**Comment by CyrusNajmabadi** [comment]
@rchande The issue is that skipping is very scenario specific.  You only want to skip for certain types of messages, when you know it's totally safe to do so.  For example, you can't skip a retrigger that is based on a typed-char, because the next  retrigger might be for a different character which may lead to different results.

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs]: sure.

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs]: Fixed.

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller.Session_ComputeModel.cs]: @fixed.

**Comment by Pilchie** [comment]
Finally gave this a try with release bits today and it definitely feels much better!  Thanks @CyrusNajmabadi

**Pull request merged**